torch/utils/data/_utils/collate.py

```diff
@@ -26,11 +26,19 @@ def default_convert(data):
             return data
         return torch.as_tensor(data)
     elif isinstance(data, collections.abc.Mapping):
-        return {key: default_convert(data[key]) for key in data}
+        try:
+            return elem_type((key, default_convert(data[key])) for key in data)
+        except BaseException:
+            # The mapping type may not support `__init__(iterable)`.
+            return {key: default_convert(data[key]) for key in data}
     elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
         return elem_type(*(default_convert(d) for d in data))
     elif isinstance(data, collections.abc.Sequence) and not isinstance(data, string_classes):
-        return [default_convert(d) for d in data]
+        try:
+            return elem_type(default_convert(d) for d in data)
```

Comment:
Okay, but I'm not entirely convinced about this, because are creating extra memory.

Reply:
Similarly for mapping? Should we create a dict instead of an iterable of tuples?

Reply:
Correct me if I am wrong. I don't think it will create extra memory. No matter if it's generator or list, it will create O(N) space eventually. It depends on how users handle the input. If they do an extra copy, then yes, it will create extra memory. If it's just reference, then no extra space.

> Similarly for mapping? Should we create a dict instead of an iterable of tuples?

Yes, please. Thank you.

Reply:
It will create an extra copy, yes. E.g., if `elem_type` == `list`, then `list([1, 2, 3])` would first create the list `[1, 2, 3]`, then create it again with new memory.

I understand that in 99% of the cases it doesn't matter. However, when using a giant batch size, where each instance takes up considerable memory, it's gonna add up significantly.

Reply:
This would be the case for `convert` and `pin_memory`. For collate, the extra memory is required anyways.

The main concern is still that the user's CustomList would accept generator object or not. If not, the previous implementation can be annoying for users.

I would say if users are worried about mem copy, they should do non-copy creation for their CustomList.

Thanks for showing the concern. Will import the PR.